spec/formatting.md

```diff
@@ -600,227 +608,87 @@ First, resolve the values of each _selector_:
-   1. Set `i` to be `i` - 1.
-1. Let `var` be the _variant_ element of the first element of `sortable`.
-1. Select the _pattern_ of `var`.
-
-`SortVariants` is a method whose single argument is
-a list of (integer, _variant_) tuples.
-It returns a list of (integer, _variant_) tuples.
-Any implementation of `SortVariants` is acceptable
-as long as it satisfies the following requirements:
-
-1. Let `sortable` be an arbitrary list of (integer, _variant_) tuples.
-1. Let `sorted` be `SortVariants(sortable)`.
-1. `sorted` is the result of sorting `sortable` using the following comparator:
-   1. `(i1, v1)` <= `(i2, v2)` if and only if `i1 <= i2`.
-1. The sort is stable (pairs of tuples from `sortable` that are equal
-   in their first element have the same relative order in `sorted`).
-
-#### Pattern Selection Examples
+Next, using `res`:
 
-_This section is non-normative._
+1. Let `bestVariant` be `UNSET`.
+1. For each _variant_ `var` of the message, in source order:
+   1. Let `keys` be the _keys_ of `var`.
+   1. Let `match` be SelectorsMatch(res, keys).
+   1. If `match` is false:
+      1. Continue the loop.
+   1. If `bestVariant` is `UNSET`.
+      1. Set `bestVariant` to `var`.
+   1. Else:
+      1. Let `bestVariantKeys` be the _keys_ of `bestVariant`.
+      1. If SelectorsCompare(`res`, `keys`, `bestVariantKeys`) is `BETTER`:
+         1. Set `bestVariant` to `var`.
+1. Select the _pattern_ of `bestVariant`.
+
+#### SelectorsMatch
+
+SelectorsMatch(`selectors`, `keys`) is defined as follows, where
+`selectors` is a list of _resolved values_
+and `keys` is a list of _keys_:
+
+1. Let `i` be 0.
+1. For each key `key` in `keys`:
+   1. If `key` is the catch-all key `'*'`
+      1. Set `i` to `i` + 1.
+      1. Continue the loop.
+   1. Let `k` be NormalizeKey(`key`).
+   1. Let `sel` be the `i`th element of `selectors`.
+   1. If Match(`sel`, `k`) is false:
+      1. Return false.
+   1. Set `i` to `i` + 1.
+1. Return true.
+
+#### SelectorsCompare
+
+SelectorsCompare(`selectors`, `keys1`, `keys2`) is defined as follows, where
+`selectors` is a list of _resolved values_
+and `keys1` and `keys2` are lists of _keys_.
+
+1. Let `result` be false.
+1. Let `i` be 0.
+1. For each _key_ in `keys1`:
+   1. Let `key1` be the `i`th element of `keys1`.
+   1. Let `key2` be the `i`th element of `keys2`.
+   1. If `key1` is the catch-all _key_ `'*'` and `key2` is not the catch-all _key_:
+      1. Return false.
+   1. If `key1` is not the catch-all _key_ `'*'` and `key2` is the catch-all _key_:
+      1. Return true.
+   1. If `key1` and `key2` are both the catch-all _key_ `'*'`
+      1. Set `i` to `i + 1`.
+      1. Continue the loop.
+   1. Let `k1` be NormalizeKey(`key1`).
+   1. Let `k2` be NormalizeKey(`key2`).
+   1. Let `sel` be the `i`th element of `selectors`.
+   1. Set `result` to BetterThan(`sel`, `k1`, `k2`).
+   1. If `result` is false:
+      1. Set `i` to `i + 1`.
+      1. Continue the loop.
+   1. Else:
+      1. Return true.
```

Comment:
Without `SAME`, we only continue if the keys are equal, and we need to include that comparison here.

```suggestion
   1. If `k1` and `k2` consist of the same sequence of Unicode code points, then:
      1. Set `i` to `i + 1`.
      1. Continue the loop.
   1. Let `sel` be the `i`th element of `selectors`.
   1. Set `result` to BetterThan(`sel`, `k1`, `k2`).
   1. Return `result`.
```

Reply:
This also removes the need to have a mutable `result` in this operation.